### ISSUE TITLE
pagination occasionally cuts off last few rows of last page

### ISSUE THREAD

**Issue by ktiedt** [angular-ui/ng-grid#3209]
I realize this is kind of a bad bug report since I am unable to consistently make this happen outside my application, however I have tracked down the code that is causing the problem and am hoping that someone more in touch with that particular code can maybe offer an explanation or better thoughts on why this is happening.

The problematic line of code is here: (snippet below) https://github.com/angular-ui/ng-grid/blob/58a27fc5bb798af3cae57be202bfed5017a6f191/src/js/core/factories/GridRenderContainer.js#L423

```
 rangeEnd = Math.min(rowCache.length, rowIndex + minRows + self.grid.options.excessRows);
```

which equates in our case:
```
 rangeEnd = Math.min(25, [6 || 9] + 10 + 4);
```
This causes our final page to either display 20 or 23 rows... cutting off the last 2-5 rows out of our 200..

Our temporary work around was to increase excessRows to ensure that 25 was always the minimum value... 

Alternatively we also found that these options worked to "fix" the problem too:
 disabling virtualization  (set its value to our max row count)
increase virtualization threshold to between 1 and 2 times our page size (25-50)

**Comment by ktiedt** [comment]
@derrabus would you be able to shed some light on the workings of this?

**Comment by PaulL1** [comment]
@ktiedt: is this still occurring?  There we substantial changes went in on the scrolling, and I think the issue here is more scrolling related than pagination related.

**Comment by ktiedt** [comment]
@PaulL1 I worked around it via the methods I mentioned in my initial post, despite not believing that to be the best fix. I can remove my fix and do a check using latest unstable. I had primarily been waiting on confirmation of what I found being the problem or not. Based on your comment, scrolling could definitely be what was causing the offset error. I'll be able to test this later this afternoon.

**Comment by ktiedt** [comment]
@PaulL1 actually was able to take care of this sooner than I expected, I still see fairly consistent rendering as previously... anywhere from 195-198 is my last index visible out of my 200 row data set.

**Comment by PaulL1** [comment]
Any chance of a plunker that would recreate it?  I've had my eye on this one for a while but cannot get a recreate that would let me look at it.  I have a theory that the logic that derives the currently visible scrolling window is a bit broken, but I haven't really taken the time to go through it from first principles.

**Comment by ktiedt** [comment]
I have tried to no avail to produce it in a plunkr, that is why I resorted to turning it into a math problem to see where it was failing for us. 

I know, that the typical response is, 'then its not a bug', perhaps the only thing I have not tried, is isolating the CSS for the grid, however, that is a monumental task in this project (to disable any non ui-grid CSS that may be throwing off calculations).

My team is 'ok' with any of the fixes we found (adjusting the options) but it would be amazing to nail down the exact cause of it and remove the need for a bandaid.

**Comment by PaulL1** [comment]
Yeah, my feel is that it probably happens for one of two reasons:

1.  There is a "you didn't scroll enough" check, usually defaults to 4 rows.  It exits if we didn't hit that threshold.  Maybe pagination is messing with that logic a bit
2.  I think there is a debounce hiding in here, so maybe it never gets to the bottom because the last debounce never runs or something.
